Add regression issue template (#77725)
rust-lang/rust: .github/ISSUE_TEMPLATE/regression.md

---
name: Regression

> about: Report a regression in Rust.

Reviewer: ```suggestion
about: Report a regression the Rust compiler or `rustdoc`.
```

Reviewer 2: I think there's something missing in this suggestion

Author: I just went off of what was there for "Bug Report":

![image](https://user-images.githubusercontent.com/37223377/95520221-a0e63180-097b-11eb-8579-08f568da8e42.png)

Should it be changed?

Reviewer: Oh hmm if that's what it is on the other issue template no need to change it here.

labels: C-bug
---
<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

### Code

I tried this code:

```rust
<code>
```

> 

Reviewer: ```suggestion
I expected to see this happen: *explanation*

Instead, this happened: *explanation*

```

### Version it worked on

<!--
Provide the most recent version this worked on, for example:

It most recently worked on: Rust 1.47
-->

It most recently worked on: <!-- version -->

### Version with regression

<!--
Provide the version you are using that has the regression.
-->

`rustc --version --verbose`:
```
<version>
```

<!--
Did the compiler crash? If so, please provide a backtrace.
-->

### Backtrace
<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

<!--

> Please change `nightly` to the release channel you found this regression in.

Reviewer: ```suggestion
Please change `nightly` to the release channel you found this regression in (stable, beta, or nightly).
```

---
name: Regression
about: Report a regression in Rust.
labels: C-bug
---
<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

### Code

I tried this code:

```rust
<code>
```
I expected to see this happen: *explanation*

Instead, this happened: *explanation*

### Version it worked on

<!--
Provide the most recent version this worked on, for example:

It most recently worked on: Rust 1.47
-->

It most recently worked on: <!-- version -->

### Version with regression

<!--
Provide the version you are using that has the regression.
-->

`rustc --version --verbose`:
```
<version>
```

<!--
Did the compiler crash? If so, please provide a backtrace.
-->

### Backtrace
<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

<!--
Please change `nightly` to the release channel you found this regression in (stable, beta, or nightly).
-->

> @rustbot modify labels: regression-from-stable-to-nightly

Reviewer 3: Hm, I wonder if this will turn out to be a bad idea. Maybe we can have the default be something like "regression-untriaged" (I can create the label).

I worry that we'll otherwise get people filing nightly regressions, which are actually beta regressions.

Author: > Maybe we can have the default be something like "regression-untriaged" (I can create the label).

That seems like a good idea 👍 

> I worry that we'll otherwise get people filing nightly regressions, which are actually beta regressions.

Well it would probably be discovered when it's triaged, but yeah it probably makes sense to default to `regression-untriaged` so there's no confusion.

Author: Maybe we have the default be `regression-untriaged` and then say in a comment that you can change it if you know where it occurred?

Reviewer 3: Yes, that seems reasonable. If it really becomes a problem we'll just remove the second bit.

Reviewer 4: By the way, I've noticed as @Mark-Simulacrum points out that from time to time we've things tagged as nightly regressions that are beta regressions. As part of @rust-lang/wg-prioritization effort, I've been checking the ones that we track for the compiler weekly meeting.

I wonder if we could make a process on Triage WG and/or Prioritization WG and even automate some stuff to more actively check that the regressions are correctly labelled.

---
name: Regression

> about: Report a regression in Rust.

Reviewer 5: What about
```suggestion
about: Report a bug for something that used to work but doesn't anymore.
```

Author: That seems good. What do other people think?

Reviewer 6: left a [comment](https://github.com/rust-lang/rust/pull/77725#issuecomment-706756620) with a possible alternative

Author: Actually just realized that not all regressions are things that used to work but don't anymore: they can also be things that shouldn't work but do. Maybe this?

```suggestion
about: Report something that unexpectedly changed between Rust versions.
```

Cc @Mark-Simulacrum -- what do you think?

labels: C-bug regression-untriaged
---
<!--
Thank you for filing a regression report! 🐛 Please provide a short summary of the regression,

> along with any information you feel is relevant to replace it.

Reviewer: Did you mean 'replicate'?
```suggestion
along with any information you feel is relevant to replicate it.
```

Author: Oops, yes :)

> Thank you for filing a regression report! 🐛 Please provide a short summary of the regression,
> along with any information you feel is relevant to replicate it.

Reviewer 6: I like what LeSeulArtichaut suggested to explain what a regression is. Therefore I'd clarify a little bit here:

```suggestion
Thank you for filing a regression! 🐛  A regression is an error in a component that used to work but doesn't anymore.

Please provide a short summary of the regression, along with any information you feel is relevant to replicate it.
```

Author: Hmm, not *all* regressions are new errors that didn't used to occur. What do you think of this?

```suggestion
Thank you for filing a regression report! 🐛 A regression is something that changed between versions of Rust but was not supposed to.

Please provide a short summary of the regression,
along with any information you feel is relevant to replicate it.
```

Reviewer 6: Probably both more correct than "my" definition and clearer in layman's terms. :+1:

-->

### Code

I tried this code:

```rust
<code>
```

I expected to see this happen: *explanation*

Instead, this happened: *explanation*

### Version it worked on

<!--
Provide the most recent version this worked on, for example:

> It most recently worked on: Rust 1.47

Reviewer: I don't think we need both this and the text below. Maybe put 'Rust 1.47' in the `version` comment?

Author: Hmm, I think it would be confusing that way. Personally I would prefer to be a bit verbose rather than potentially unclear.

Reviewer: Ok, seems fine.

-->

It most recently worked on: <!-- version -->

### Version with regression

<!--
Provide the version you are using that has the regression.
-->

`rustc --version --verbose`:
```
<version>
```

<!--
Did the compiler crash? If so, please provide a backtrace.
-->

### Backtrace
<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

<!--
If you know when this regression occurred, please add a line like below, replacing `{channel}` with one of stable, beta, or nightly.

> @rustbot modify labels: regression-from-stable-to-{channel}

Author: Hmm, should I change it to this so that the `regression-untriaged` label isn't added in addition to `regression-from-stable-to-{channel}`?

```suggestion
@rustbot modify labels: +regression-from-stable-to-{channel} -regression-untriaged
```

Reviewer 3: That seems good, yes.
